File: crates/sdk/python/clawlegion/tool.py
"""Tool system for Python plugins."""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional
from .visibility import Visibility


@dataclass
class ToolMetadata:
    """Metadata for a tool."""

    name: str
    """Unique tool name."""

    version: str
    """Tool version (semver)."""

    description: str
    """Tool description."""

    visibility: Visibility = Visibility.PUBLIC
    """Tool visibility."""

    tags: List[str] = field(default_factory=list)
    """Tags for categorization and retrieval."""

    input_schema: Dict[str, Any] = field(default_factory=lambda: {"type": "object"})
    """Input JSON schema."""

    output_schema: Optional[Dict[str, Any]] = None
    """Output JSON schema."""

    requires_llm: bool = False
    """Whether this tool requires LLM."""


@dataclass
class ToolContext:
    """Context provided during tool execution."""

    agent_id: str
    """Agent ID executing this tool."""

    config: Dict[str, Any] = field(default_factory=dict)
    """Tool configuration."""

    timeout_ms: Optional[int] = None
    """Execution timeout in milliseconds."""

    def get_config(self, key: str, default: Any = None) -> Any:
        """Get a configuration value."""
        return self.config.get(key, default)


@dataclass
class ToolResult:
    """Result from tool execution."""

    success: bool = True
    """Success flag."""

    data: Optional[Any] = None
    """Result data."""

    error: Optional[str] = None
    """Error message if failed."""

    execution_time_ms: int = 0
    """Execution time in milliseconds."""

    @classmethod
    def success(cls, data: Any) -> "ToolResult":
        """Create a successful result."""
        return cls(data=data, success=True)

    @classmethod
    def error(cls, message: str, execution_time_ms: int = 0) -> "ToolResult":
        """Create an error result."""
        return cls(success=False, error=message, execution_time_ms=execution_time_ms)


class Tool(ABC):
    """Base class for all ClawLegion tools."""

    @abstractmethod
    def metadata(self) -> ToolMetadata:
        """Get tool metadata."""
        pass

    @abstractmethod
    async def execute(self, ctx: ToolContext, args: Dict[str, Any]) -> ToolResult:
        """Execute the tool.

        Args:
            ctx: Tool context with agent info and configuration.
            args: Tool arguments (must match input_schema).

        Returns:
            ToolResult: Result of tool execution.

        Raises:
            Exception: If execution fails.
        """
        pass

    def description(self) -> str:
        """Get tool description for LLM prompts.

        Returns:
            str: Tool description.
        """
        return self.metadata().description

    async def validate_args(self, args: Dict[str, Any]) -> bool:
        """Validate tool arguments against input schema.

        Args:
            args: Arguments to validate.

        Returns:
            bool: True if valid, False otherwise.
        """
        # Simple validation - can be extended with jsonschema
        return isinstance(args, dict)

File: crates/capability/src/tool.rs
//! Tool system - basic tool functions

use async_trait::async_trait;
use clawlegion_core::{AgentId, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;

/// Tool visibility
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Visibility {
    /// Public tools are available to all agents
    Public,

    /// Private tools must be bound to a Skill
    Private,
}

/// Tool metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolMetadata {
    /// Unique tool name
    pub name: String,

    /// Tool version (semver)
    pub version: String,

    /// Tool description
    pub description: String,

    /// Visibility
    pub visibility: Visibility,

    /// Tags for categorization and retrieval
    pub tags: Vec<String>,

    /// Input schema (JSON Schema)
    pub input_schema: serde_json::Value,

    /// Output schema (JSON Schema)
    pub output_schema: Option<serde_json::Value>,

    /// Whether this tool requires LLM
    pub requires_llm: bool,
}

/// Tool execution context
#[derive(Debug, Clone)]
pub struct ToolContext {
    /// Agent ID executing this tool
    pub agent_id: AgentId,

    /// Tool configuration
    pub config: HashMap<String, serde_json::Value>,

    /// Execution timeout in milliseconds
    pub timeout_ms: Option<u64>,
}

impl ToolContext {
    pub fn new(agent_id: AgentId) -> Self {
        Self {
            agent_id,
            config: HashMap::new(),
            timeout_ms: None,
        }
    }

    pub fn with_config(mut self, config: HashMap<String, serde_json::Value>) -> Self {
        self.config = config;
        self
    }

    pub fn with_timeout(mut self, timeout_ms: u64) -> Self {
        self.timeout_ms = Some(timeout_ms);
        self
    }
}

/// Tool execution result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolResult {
    /// Success flag
    pub success: bool,

    /// Result data
    pub data: Option<serde_json::Value>,

    /// Error message if failed
    pub error: Option<String>,

    /// Execution time in milliseconds
    pub execution_time_ms: u64,
}

impl ToolResult {
    pub fn success(data: serde_json::Value) -> Self {
        Self {
            success: true,
            data: Some(data),
            error: None,
            execution_time_ms: 0,
        }
    }

    pub fn error(message: impl Into<String>) -> Self {
        Self {
            success: false,
            data: None,
            error: Some(message.into()),
            execution_time_ms: 0,
        }
    }
}

/// Tool trait - defines basic tool functions
#[async_trait]
pub trait Tool: Send + Sync {
    /// Get tool metadata
    fn metadata(&self) -> &ToolMetadata;

    /// Execute the tool with given arguments
    async fn execute(&self, ctx: &ToolContext, args: serde_json::Value) -> Result<ToolResult>;

    /// Get a description of what this tool does (for LLM prompts)
    fn description(&self) -> &str {
        &self.metadata().description
    }
}

/// Type-erased tool wrapper
pub type ToolBox = Arc<dyn Tool>;

/// Tool registry entry
pub struct ToolEntry {
    pub tool: ToolBox,
    pub bound_skills: Vec<String>, // Skills that have bound this private tool
}

impl ToolEntry {
    pub fn new(tool: ToolBox) -> Self {
        Self {
            tool,
            bound_skills: vec![],
        }
    }

    pub fn bind_to_skill(&mut self, skill_name: &str) {
        if !self.bound_skills.iter().any(|s| s == skill_name) {
            self.bound_skills.push(skill_name.to_string());
        }
    }

    pub fn is_accessible_by_skill(&self, skill_name: &str) -> bool {
        self.tool.metadata().visibility == Visibility::Public
            || self.bound_skills.iter().any(|s| s == skill_name)
    }
}

/// Macro for defining tools
#[macro_export]
macro_rules! define_tool {
    (
        name = $name:literal,
        version = $version:literal,
        description = $description:literal,
        visibility = $visibility:ident,
        tags = [$($tag:literal),*],
        input_schema = $input_schema:expr,
    ) => {
        pub struct ToolDef;

        impl ToolDef {
            pub fn metadata() -> $crate::ToolMetadata {
                $crate::ToolMetadata {
                    name: $name.to_string(),
                    version: $version.to_string(),
                    description: $description.to_string(),
                    visibility: $crate::Visibility::$visibility,
                    tags: vec![$($tag.to_string()),*],
                    input_schema: $input_schema,
                    output_schema: None,
                    requires_llm: false,
                }
            }
        }
    };
}
